### DISCUSSION TITLE
Add `PlayerSpawnInUnknownWorldEvent`

### DISCUSSION THREAD

**Discussion by molor** [PaperMC/Paper#10571] [Ideas]
### Is your feature request related to a problem?

Yea, our players sometimes gets stuck in blocks (in main overworld at same position where they were in the other world) and then die due to suffocation :/

This happens only if the server was restarted and the other world's folder (created by plugin or vanilla nether/the end) was deleted.

### Describe the solution you'd like.

Add `PlayerSpawnInUnknownWorldEvent` (or something with better name) that gets called if player joining the game and the server can't find the world this player was in before.

That event contains `#setSpawnLocation(Location)` and nullable `#getSpawnLocation()`. By default that `Location` is `null` to keep existing behavior. If no one plugin changed this `Location`, then this player will spawn in overworld at same position (current behavior), otherwise the server will use that `Location` to spawn this player at.

### Describe alternatives you've considered.

idk. I'm tried `PlayerSpawnLocationEvent` but it gets called with default world/overworld in this case, so plugins cannot determine exactly which world the player was in before.

**Comment by Doc94** [comment]
Not sure about a event like this to "solve a issue"

**Comment by molor** [comment]
Something like `move-from-unknown-worlds-to-overworld-spawn` in `paper-global.yml` could also be a solution.

**Comment by Malfrador** [comment]
This seems like something that could be easily fixed by just having the plugin keep track of the last world a player was in itself (using the PlayerChangedWorldEvent and some data storage).

So I'm not really convinced that this needs to be an API addition. A config option on the other hand could be useful to avoid needing a plugin specifically for this.

**Comment by molor** [comment]
> This seems like something that could be easily fixed by just having the plugin keep track of the last world a player was in itself (using the PlayerChangedWorldEvent and some data storage).

A little later I also thought about it, but we already have a lot of players who quit in worlds that no longer exist. We don't know their names, but we know that they are exist.. So this "fix" will work for "new" players only :(

> A config option on the other hand could be useful to avoid needing a plugin specifically for this.

Or just tp the player to overworld spawn by default, instead of changing his world and keeping at the same coordinates. It seems to me more logical, safe (for player) and predictable..

**Comment by molor** [comment]
Another alternative solution: Add nullable `getLastWorld()` to `OfflinePlayer`, so if it return `null` during `AsyncPlayerPreLoginEvent` then plugin can teleport the player to overworld spawn using `PlayerSpawnLocationEvent`.

**Comment by Jikoo** [comment]
Paper already has a patch that is supposed to take care of this, is it not working? https://github.com/PaperMC/Paper/blob/f8c01122e69ce1f7681e6bc2b8e280ce807167df/patches/server/0384-Move-player-to-spawn-point-if-spawn-in-unloaded-worl.patch

**Comment by molor** [comment]
@Jikoo, it doesn't move player to spawn **point**, it move player to the spawn **world** to the same position he was in unloaded world. That position may contain lava or may be located under ground, and player will just die for example due to suffocation. There's currently no way to avoid/control this.

**Comment by Jikoo** [comment]
Then the patch isn't working. It's supposed to move the player to their spawn if available or the default spawn if the world they're in is nonexistent. More accurately, the patch specifically just leaves it to MC to figure out what to do about a null world instead of doing the Craftbukkit thing and moving them to the default world at their current coordinates. Maybe default MC behavior has changed.

**Comment by pop4959** [comment]
It does appear the description of that patch does not match the behavior.

If that bug is fixed would you still need this API? I feel like it will become rather irrelevant then.

**Comment by molor** [comment]
> If that bug is fixed would you still need this API?

If the question is for me, then after all that I wrote earlier, I don't quite understand what bug you are talking about.. Everything that is described [here](https://github.com/PaperMC/Paper/issues/9081#issuecomment-1501147490) is still a problem, at least for me.

**Comment by Doc94** [comment]
Hmm with https://github.com/PaperMC/Paper/blob/master/patches/server/0373-Don-t-move-existing-players-to-world-spawn.patch and https://github.com/PaperMC/Paper/blob/master/patches/server/0384-Move-player-to-spawn-point-if-spawn-in-unloaded-worl.patch maybe i understand what happen.... i test editing the last patch for handle the method used in the first patch for handle the spawn position only when the world was change (in teory when the world its not found) if works then can open a PR...

**Comment by imDaniX** [comment]
> This seems like something that could be easily fixed by just having the plugin keep track of the last world a player was in itself (using the PlayerChangedWorldEvent and some data storage).

I don't think so. In case a plugin was installed after deleting a world, it just won't know what world a player was in last time.
Also, to me, the existence of such event would just make sense in terms of handling player movement. Not everyone wants a player to be teleported to default world spawn - e.g. home location might be desired instead.

**Comment by Machine-Maker** [comment]
I think an event for this is a good idea, and once https://github.com/PaperMC/Paper/pull/9887 is merged, an event will be easy to add on top of it.
